---
title: "Reproducible research"
author: "Albert Cobos"
output: 
  ioslides_presentation:
    widescreen: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      warning = FALSE, 
                      message = FALSE)

# packages
library(dplyr)
library(ggformula)
library(patchwork)


```

## Reproducibility of scientific research

Concerns about lack of reproducibility of scientific research ([Baker, Nature 2016](https://www.nature.com/articles/533452a)).

```{r }
knitr::include_graphics("./images/reproducibility_crisis.PNG", dpi = 120)
knitr::include_graphics("./images/reproducibility_causes.PNG", dpi = 120)
```

## Computational reproducibility (lack of) {.smaller}

Reproducibility of results when the original data is available.

- [Ioannidis, Nature Genetics 2009](https://www.nature.com/articles/ng.295):

    > Evaluated 18 articles on microarray gene expression profiling published in Nature Genetics in 2005â€“2006. We reproduced 2, and 6 partially or with some discrepancies; **10 could not be reproduced**.
    
    > Failure to reproduce were mostly due to incomplete data annotation or **incomplete specification of data processing and analysis**.

- [Hardwike et al., RSOS 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6124055/))

    > For 22/35 articles all values were reproduced, but 11 required author assistance.
    
    > For 13/35 articles at least one value could not be reproduced **despite author assistance**.
    
    > Suboptimal data curation, **unclear analysis specification and reporting errors** can impede analytic reproducibility, undermining the utility of data sharing and the credibility of scientific findings. 



## Literate programming

A single document (file) that combines **text** and **R code**.

\

R packages to help with LP:

- `knitr`: report generation from combined text and R code.

- `Rmarkdown`: simple formatting syntax for text and code.

- `tinytext`:  needed to create pdf documents.

\

You should install `tinytext` now !


## Creating a report

In RStudio:

1. `File > New file > R mardown...`

2. A dialog will open, where you should:
    + Provide a title and author in the dialog  
    + Leave the default output format (HTML)  
    + Click OK
    
3. When a new `Untitled1` tab opens, click the `Knit` button. 

4. Give a name to the file: it **should not** containg blank spaces!

5. A new HTML report should appear in a new window.

6. This is an HTML file that has been added to your working directory.


## Report output formats

- HTML

- pdf

- MS Word

\

Unfold the `knit` button to pick the desired one.


## Rmarkdown document (.Rmd) 

Structure of an .Rmd document:

- YAML header, delimited by three dashes (`---`)

- Text, with R markdown formatting 

- Code chunks for R code, with optional name and chunk options

## YAML header

Some possibilities (some of them do not work in some output formats):

````
---
title: "Birth weight"
author: "Albert Cobos"
date: "04/08/2021"
output:
  pdf_document:
    toc: true
    toc_depth: 2
---
````

For more YAML options, see [here](https://bookdown.org/yihui/rmarkdown/html-document.html)


## Text, with R markdown formatting 

- Use hashes (`##`) for section headers: the number of hashes determine the level of header.  

- `**boldface**` or `__boldface__` become **boldface**  

- `*italics*`  or `_italics_` become *italics*  

- `<https://cdb.clinic.cat/>` becomes <https://cdb.clinic.cat/>  

- `[Clinic CDB](https://cdb.clinic.cat/)` becomes [Clinic CDB](https://cdb.clinic.cat/)

- ...


For more R markdown formatting options, see:

- the [R Markdown reference guide](https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf?_ga=2.31117371.589981273.1627977413-1000639518.1610891435), or 

- the [R Markdown cheatsheet](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf)


## Chunk options

Options affecting text output of R code:

- `echo = FALSE`: to suppress the R code (default is `TRUE`, showing the R code).

- `message = FALSE`: to suppress R messages (default is `TRUE`, showing messages).

- `warning = FALSE`: to suppress R warnings (default is `TRUE`, showing warnings).

- `error = FALSE`: to suppressed R errors (default is `TRUE`, showing errors).

- `results = 'asis'`: to write text results unaltered (default is `markup`, writing text according to output format).


## Chunk options

Options affecting plots:

- `fig.align = 'center'`: To center figures (default is `'default'`, which does not make any alignment adjustment; other possible values are `'left'` or `'right'`).

- `fig.cap = 'A caption for the figure'`: To provide, within quotes, a caption for the figure (default is `NULL`). Figures with a caption will be automatically numbered, and can be referenced in the text (in pdf output documents).

- `fig.height = ` _n_, `fig.width = ` _n_: Set the figure height and width to the number of inches provided (_n_).


## Chunk options

- Can be specified in any particular code chunk, and then they will be applied to this chunk only, other chunks remaining unaffected

- Can be set with function `knitr::opts_chunk$set()` in the first code chunk, an then they will affect all subsequent code chunks (unless overwritten in a particular code chunk)

Useful first code chunk:

````
```{r setup, include = FALSE}`r ''`
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE,
                      comment=NA, fig.align = 'center')
                     
```
````

\

For more chunk options, see [here](https://yihui.org/knitr/options/)

## Slides presentation

R Markdown can be used to produce slide presentations in several output formats:

- HTML: IOslides, and Slidy

- pdf:  Beamer

- MS PowerPoint


## Creating an IOslides presentation {.smaller}

In RStudio:

1. `File > New File > R Markdown...`

2. A dialog will open, where you should:
    + Choose `Presentation` in the left panel
    + Provide a title and author in the dialog  
    + Leave the default output format HTML(ioslides)  
    + Click OK
    
3. When a new `Untitled1` tab opens, click the `Knit` button. 

4. Give a name to the file: it **should not** containg blank spaces!

5. A new HTML presentation should appear in a new window.

6. This is an HTML file that has been added to your working directory.


## Exercise

1. Create a .pdf report on the birth weight data in `MASS::birthwt`:

    + In the first setup chunk:
        + define general chunk options
        + load the packages you need with `library()`
        + read and prepare the data (see code in next slide)
    + Write an **Introduction** section with some text.
    + Write a **Methods** section with some text.
    + Write a **Results** section with at least two graphics of your choice, and some sentences interpreting the graphics.


2. Create an IOslides presentation with similar contents as exercise 1.

## Get and prepare the `MASS::birthwt` dataset

```{r echo = TRUE, comment = NA}
d <- MASS::birthwt %>%
  mutate(low = factor(low, levels = 1:0, labels = c("low", "normal")),
         race = factor(race, levels = 1:3, labels = c("white", "black", "other")),
         smoke = factor(smoke, levels = 1:0, labels = c("smoker", "non-smoker")),
         ui = factor(ui, levels = 1:0, labels = c("UI", "no-UI")),
         ht = factor(ht, levels = 1:0, labels = c("AHT", "no-AHT"))) %>% 
  rename(bw = low)

head(d)
```

